### PR TITLE
DPC-1437 Feature: Add new POST /v2/introspect route to validate token

### PR DESCRIPTION
**Pull request by charmcitygavin**
### Fixes [DPC-1437](https://jira.cms.gov/browse/DPC-1437)

We needed a new `POST /v2/introspect` endpoint that simply returns whether or not a token is valid. This request will be proxied from `dpc-app`.

### Proposed Changes

* Create a new `POST /v2/introspect` route (to conform to the [tech specs](https://confluence.cms.gov/display/DAPC/DPC+Version+2+Technical+Specification))
* Return `"valid": true` or `"valid": false` depending on whether or not the token is valid
* Return information to the user if they are missing required claims

### Change Details

* Added the new route
* Return a response body with true or false depending on whether or not the token is valid
* Return an error string with information about any missing claims in the JWT

### Security Implications

- [ ] new software dependencies

- [ ] security controls or supporting software altered

- [ ] new data stored or transmitted

- [x] [security checklist](https://confluence.cms.gov/display/DAPC/2021-09-02%3A+DPC+4.2+Security+Checklist) is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

- [x] no PHI/PII is affected by this change

### Acceptance Validation

Yes, in Postman and with unit tests:

Screen Shot 2021-09-09 at 09 38 40

Screen Shot 2021-09-09 at 09 37 47

Screen Shot 2021-09-09 at 09 36 52

This PR creates a new endpoint and doesn't interfere with existing client APIs.

### CODE REVIEW

File: ssas/service/public/api.go [hunk 1]
```diff
@@ -764,3 +764,38 @@ func validateAndParseToken(w http.ResponseWriter, r *http.Request) {
...
+	var response = make(map[string]interface{})
+
+	if err := tokenValidity(tokenS, "AccessToken"); err != nil {
+		ssas.Logger.Infof("token failed tokenValidity")
+		response["valid"] = false
+	} else {
+		claims := jwt.MapClaims{}
+		if _, _, err := new(jwt.Parser).ParseUnverified(tokenS, claims); err != nil {
+			ssas.Logger.Infof("could not unmarshal access token")
+			jsonError(w, http.StatusText(http.StatusInternalServerError), "internal server error")
```
Comment (jonfulk) [inline ssas/service/public/api.go:794]: is this actually an internal server error or something else?

**Review by jonfulk — APPROVED** [review]
looks simple and straightforward to me!

**Comment by jonfulk** [comment]
i should have mentioned that my approval is based on the assumption that we have this ticket in a security checklist already and that you'll be linking to it here.

**Comment by charmcitygavin** [comment]
Thanks anyway, @jonfulk! I'll have it updated today and will ping you for a reapproval 👍🏻

File: ssas/service/public/api.go [hunk 2]
```diff
@@ -771,3 +771,55 @@ func validateAndParseToken(w http.ResponseWriter, r *http.Request) {
...
+	tokenS := reqBody["token"]
+	if tokenS == "" {
+		jsonError(w, http.StatusText(http.StatusBadRequest), `Missing "token" field in body`)
+		return
+	}
+
+	// parse jwt claims
+	claims := jwt.MapClaims{}
+
+	if _, _, err := new(jwt.Parser).ParseUnverified(tokenS, claims); err != nil {
```
Comment (MrBilnon) [inline ssas/service/public/api.go:798]: shouldn't you be calling system.VerifyClientSignedToken?

Comment (charmcitygavin) [reply ssas/service/public/api.go:798]: But now that I've gone back and re-read the original ticket I'm not 100% sure anymore. @dhgreene, do you know if we need to verify that the `system_id` exists in the DB at this endpoint, or is it sufficient to check that the required claims exist and give the user feedback on what's missing (if anything)?

**Comment by kyeah** [comment]
Closing as outdated. The original ticket is marked done and SSAS v2 is not expected to be in use now or in the future.

**Pull request closed without merging**